reposerver/repository/repository.go

```diff
@@ -754,8 +754,73 @@ func (s *Service) runManifestGenAsync(ctx context.Context, repoRoot, commitSHA,
 						}
 					}
 				}
-			}
-		}
+			} // if q.ApplicationSource.Helm != nil {
+
+			// Before generating the manifests, copy any structure from other
+			// This is done for all types of sources, not just helm.
+
+			// If the applicationsource has declared any subtrees to be copied
+			for _, copySpec := range q.ApplicationSource.From {
+				// First resolve the referenced source repo
+				// ref is of type RefTarget
+
+				refVar := copySpec.Ref
+				refSourceMapping, ok := q.RefSources[copySpec.Ref]
+				if !ok {
+					if len(q.RefSources) == 0 {
+						ch.errCh <- fmt.Errorf("source in from referenced %q, but no source has a 'ref' field defined", refVar)
+					}
+					refKeys := make([]string, 0)
+					for refKey := range q.RefSources {
+						refKeys = append(refKeys, refKey)
+					}
+					ch.errCh <- fmt.Errorf("source in from referenced %q, which is not one of the available sources (%s)", refVar, strings.Join(refKeys, ", "))
+					return
+				}
+				// Check for this limitation is taken from above
+				if refSourceMapping.Chart != "" {
+					ch.errCh <- fmt.Errorf("source has a 'chart' field defined, but Helm charts are not yet not supported for 'ref' sources")
+					return
+				}
+
+				// The ref is available, now - checking the tree out
+				normalizedRepoURL := git.NormalizeGitURL(refSourceMapping.Repo.Repo)
+				repoRef, ok := repoRefs[normalizedRepoURL]
+				if ok {
+					if repoRef.revision != refSourceMapping.TargetRevision {
+						ch.errCh <- fmt.Errorf("cannot reference multiple revisions for the same repository (%s references %q while %s references %q)", refVar, refSourceMapping.TargetRevision, repoRef.key, repoRef.revision)
```

Comment:
We should be able to drop this limitation for `from`, since we're making a copy of the repo instead of just locking the existing copy.

Reply:
+1

Reply:
@crenshaw-dev Commented it out for now, if it turns out it's actually needed.

@ishitasequeira I've added a check on that, if it's the same ref as the current appsource, errors out.

Reply:
I don't think we are handling a scenario in the code where we would want to reference different revisions of the same Source. Are we? For example, consider the below yaml where we are referencing same source with different versions.

```
sources:
   - repoURL: http://my/repo.git
      targetRevision: master
      ref: valuesrepo
   - repoURL: http://my/repo.git
      targetRevision: v1.0.0
      from:
	 - ref: $valuesrepo
	    sourcePath: "dev/values.yaml"
            destinationPath: "env-values.yaml"
```

Reply:
Definitely not. Error is:
```
rpc error: code = Unknown desc = cannot reference a different revision of the same repository ($envspec references "dev" which resolves to "b78be0df1c270edd0ef0ff065680f66b246fa604" while the application references "main" which resolves to "55dfb7b584f3fe9f6c099c0880859e0105957c17")
```

Test repo is https://github.com/gczuczy/argocd-copyfrom-test
dev for env-spec, master for the kustomize

Since the lock is released, this shouldn't be a problem, right? Especially if different revisions are checked out to different directories.

Although I think this might be a valid usecase for some people.

Reply:
I've disabled checking for the same repo at `repository.go:816`, and it didn't go quite well. For this to work, separate checkout is needed per repo+revision.

I'm leaning forward adding a this to the limitations of the implementation for now, included in the documentation, and lettings this be specific case implemented in a separate PR. This way it could be seen how much need is there among the userbase - despite seeing a relatively valid usecase in it, it doesn't mean people are actually willing to form their repositories this way actually.